mlir/include/mlir/Dialect/MQTOpt/Transforms/Transpilation/Planner.h

```diff
@@ -0,0 +1,254 @@
+    [[nodiscard]] mlir::SmallVector<QubitIndexPair> getSequence() const {
+      return seq_;
+    }
+
+    /**
+     * @brief Return a const reference to the node's layout.
+     */
+    [[nodiscard]] const ThinLayout& getLayout() const { return layout_; }
+
+    [[nodiscard]] bool operator>(const SearchNode& rhs) const {
+      return f_ > rhs.f_;
+    }
+
+  private:
+    /**
+     * @brief Calculate the path cost.
+     *
+     * The path cost function evaluates the weighted sum of the currently
+     * required SWAPs and additionally added depth.
+     */
+    [[nodiscard]] float g(const HeuristicWeights& weights) {
+      return (weights.alpha * static_cast<float>(seq_.size())) +
+             (weights.beta * static_cast<float>(ndepth_));
+    }
+
+    /**
+     * @brief Calculate heuristic cost.
+     *
+     * The heuristic cost function calculates the nearest neighbour costs.
+     * That is, the amount of SWAPs that a naive router would require.
+     * Gamma acts like a decay.
+     *
+     * TODO: Similarly to the LightSABRE algorithm. It should be possible to
+     * calculate the heuristic in O(1) by only considering the change that the
+     * inserted SWAP causes.
+     */
+    [[nodiscard]] float h(const Layers& layers, const Architecture& arch,
+                          const HeuristicWeights& weights) {
+      float nn{0};
+      for (const auto [i, layer] : llvm::enumerate(layers)) {
+        for (const auto [prog0, prog1] : layer) {
+          const auto [hw0, hw1] = layout_.getHardwareIndices(prog0, prog1);
+          const std::size_t dist = arch.distanceBetween(hw0, hw1);
+          const std::size_t nswaps = dist < 2 ? 0 : dist - 2;
+          nn += weights.lambdas[i] * static_cast<float>(nswaps);
+        }
+      }
+      return nn;
+    }
+
+    mlir::SmallVector<QubitIndexPair> seq_;
+    ThinLayout layout_;
+
+    mlir::SmallVector<uint16_t> depthBuckets_;
+    uint16_t ndepth_{0};
+
+    float f_{0};
+  };
+
+  using MinQueue =
+      std::priority_queue<SearchNode, std::vector<SearchNode>, std::greater<>>;
+
+  /**
+   * @brief Expand frontier with all possible neighbouring SWAPs in the current
+   * front.
+   */
+  void expand(MinQueue& frontier, const SearchNode& node, const Layers& layers,
+              const Architecture& arch) const {
+    llvm::SmallDenseSet<QubitIndexPair, 16> visited{};
+    for (const QubitIndexPair gate : layers.front()) {
+      for (const auto prog : {gate.first, gate.second}) {
+        const auto hw0 = node.getLayout().getHardwareIndex(prog);
+        for (const auto hw1 : arch.neighboursOf(hw0)) {
+          /// Ensure consistent hashing/comparison
+          const QubitIndexPair swap =
+              hw0 < hw1 ? QubitIndexPair{hw0, hw1} : QubitIndexPair{hw1, hw0};
+
+          if (visited.contains(swap)) {
+            continue;
+          }
+
+          frontier.emplace(node, swap, layers, arch, weights_);
+          visited.insert(swap);
+        }
+      }
+    }
+  }
```

Comment:
I would expect this to be an area of the code where some improvements could be made by caching more aggressively across calls to `expand` given how the architecture information is fully static.
Nothing for this PR (obviously). just wanted to point it out for a follow up.